src/Microsoft.Extensions.Configuration/ConfigurationRoot.cs

```diff
@@ -24,6 +24,12 @@ public ConfigurationRoot(IList<IConfigurationProvider> providers)
             _providers = providers;
         }
 
+        public string Key => string.Empty;
+
+        public string Path => string.Empty;
+
+        public string Value { get; set; }
```

Comment:
Everything you have listed so far sounds more like implementation issues as opposed to real limitations. I think you identified both the problem and the solution already. The root should not be special. As you pointed out everything works if it's implementation of GetSection behaved the same. Fundamentally I don't understand why it's ok for a child to both be the root of a subtree and have a value. But somehow the root is magically different. It just smells :). To me these all feel like flaws in the current implementation which when fixed most special cases go away and every configuration element behaves the same as you would expect in a tree.